### ISSUE TITLE
Trying to retrieve AD_ID in SDK Runtime

### ISSUE THREAD

**Issue by hellosagar** [android/privacy-sandbox-samples#28]
When I'm running the below piece of code on `onLoadSDK()` in`SandboxedSdkProvider` extended class
```
private void testAdId() {
      AdIdManager adIdManager = getContext().getSystemService(AdIdManager.class);
      adIdManager.getAdId(Runnable::run, new OutcomeReceiver<AdId, Exception>() {
         @Override
         public void onResult(@NonNull AdId adId) {
            Log.d("SDK Runtime", "SDK Runtime.testAdId onResult Adid: " + adId);
         }

         @Override
         public void onError(@NonNull Exception error) {
            Log.e("SDK Runtime", "SDK Runtime.testAdId onError " + error.getMessage());
         }
      });
   }
```

This is being printed in the console
```
2022-10-04 01:09:39.174 10282-10301 SDK Runtime                              com.google.android.sdksandbox        E  
SDK Runtime.testAdId onError Caller is not authorized to call this API. Permission was not requested.
```

**Comment by siran-work** [comment]
Thanks for reaching out to us! 

This error is displayed as the caller doesn't have the permission to call AD_ID.

Did you add following in your manifest file?

<uses-permission android:name="android.permission.ACCESS_ADSERVICES_AD_ID" />

**Comment by hellosagar** [comment]
@siran-work which permission you are referring to exactly?

**Comment by jose4gonzalez** [comment]
@hellosagar 

`<uses-permission android:name="android.permission.ACCESS_ADSERVICES_AD_ID" />`

**Comment by hellosagar** [comment]
This permission is already added to the manifest, still facing the error mentioned above

<img width="1080" alt="Screenshot 2022-10-11 at 12 03 26 PM" src="https://user-images.githubusercontent.com/50016799/195013501-db6d7801-3552-493a-a9b5-79e44ab5e094.png">

**Comment by cshmerling** [comment]
You were using the UI to install the SDK correct? That actually won't use the sdk-app (and something we are looking to clarify in our sample code). It actually just uses the `example-sdk` project to pull in the right dependencies. Can you try adding the permission to the `sdk-implementation` module and let me know if that resolves your issue?

https://github.com/android/privacy-sandbox-samples/blob/main/PrivacySandboxJava/sdk-implementation/src/main/AndroidManifest.xml

**Comment by hellosagar** [comment]
Still the same behaviour

**Comment by cshmerling** [comment]
Thanks for trying. I've tested this a bit and am able to repro. There may be an issue with how SDKs are being generated using the Canary instructions and passing permissions defined in the manifest along. I'll need to follow up with some other folks to debug this further. So no immediate answer (other than if you try using the Stable Android Studio instructions and install the sdk as an app, but going forward this method will not be supported but might be able to unblock your immediate testing), so I will follow up with you when I know more

**Comment by cshmerling** [comment]
@hellosagar we published a new version of the TiramisuPrivacySandbox SDK. Can you update to revision 8 and see if your issue is resolved?

**Comment by hellosagar** [comment]
@cshmerling sure, I'll give it a shot

**Comment by hellosagar** [comment]
@cshmerling when ARM image will be released for revision 8?

<img width="917" alt="Screenshot 2022-10-31 at 10 43 14 PM" src="https://user-images.githubusercontent.com/50016799/199068068-6f4456c4-6c91-4cc2-8b45-8e9694e06edd.png">

**Comment by hellosagar** [comment]
Sorry where can I update?

**Comment by cshmerling** [comment]
The box next to the ARM image 
![image](https://user-images.githubusercontent.com/106990149/199076497-369bd5a2-4fef-46e5-a62c-a9e8fd2f9ba3.png)

**Comment by hellosagar** [comment]
After updating the project, unable to build the project

```

**Comment by cshmerling** [comment]
Looks like there are some issues with the latest Android Studio not enabling some build features. I'll need to create a PR to update the files, but in the meantime, to unblock you:

in example-aidl-library/build.gradle, add
```
        buildFeatures {
              aidl = true
           }
```
at the end of the file

**Comment by hellosagar** [comment]
@cshmerling this error was fixed and the app is installed. 

But in an android studio, it's showing only `client-app` whereas previously it showed `sdk-app` as well in the list to install the `sdk-app` first and then `client-app`

<img width="221" alt="Screenshot 2022-11-07 at 7 28 01 PM" src="https://user-images.githubusercontent.com/50016799/200328228-ced8cd01-fefd-4a47-a318-e9204d091350.png">

So, now I've installed the client-app using the UI and since I don't where to install the SDK. On click on load SDK in the app its not loading the SDK and showing the following error message

<img width="1637" alt="Screenshot 2022-11-07 at 7 30 33 PM" src="https://user-images.githubusercontent.com/50016799/200328722-1e13b0c8-7bc4-4ae3-a5c5-09ef744b7442.png">

Can you please assist from here
Thanks

**Comment by cshmerling** [comment]
sdk-app being removed is expected, as that is not actually needed to install an SDK. SDK installation should happen automatically since you are using Canary Android Studio.

The issue is some feature flags controlling access to the runtime got introduced that we missed documenting. You'll need to run:
```
adb shell device_config put adservices global_kill_switch false
adb shell device_config put adservices disable_sdk_sandbox false

```

from a terminal before launching the client app

**Comment by hellosagar** [comment]
So, I just ran the commands that you mentioned and still SDK is not loading. Following logs is being generated

```
2022-11-08 12:25:38.080 11886-11886 studio.deploy                       installer                            E  Could not get package user id: run-as: unknown package: com.example.privacysandbox.client
2022-11-08 12:25:38.085 11886-11886 studio.deploy                       installer                            E  Could not find apks for this package: com.example.privacysandbox.client
2022-11-08 12:25:38.092 11886-11886 studio.deploy                       installer                            E  Could not get package user id: run-as: unknown package: com.example.privacysandbox.client
2022-11-08 12:25:38.098 11886-11886 studio.deploy                       installer                            E  Could not find apks for this package: com.example.privacysandbox.client
2022-11-08 12:25:38.603   553-599   VerityUtils                         system_server                        E  Failed to measure fs-verity, errno 1: /data/app/~~e0HjwHVSVKvxclDnyDXL7Q==/com.example.privacysandbox.client-cv184y6UWMy7TouSjwLxOw==/base.apk
2022-11-08 12:25:43.226 14479-14500 EGL_emulation                       com.example.privacysandbox.client    D  app_time_stats: avg=376.41ms min=1.11ms max=3776.36ms count=11
2022-11-08 12:25:43.280 14479-14479 SandboxClient                       com.example.privacysandbox.client    E  Attempting to load sandbox SDK
2022-11-08 12:25:43.383 14479-14498 SandboxClient                       com.example.privacysandbox.client    E  onLoadSdkFailure(103): null
2022-11-08 12:25:43.383 14479-14479 CompatibilityChangeReporter         com.example.privacysandbox.client    D  Compat change id reported: 147798919; UID 10163; state: ENABLED
```

and right now I'm using the following version of android studio

```
Android Studio Flamingo | 2022.2.1 Canary 6
Build #AI-222.4345.14.2221.9189903, built on October 18, 2022
Runtime version: 17.0.4.1+0-17.0.4.1b469.62-9127311 aarch64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
macOS 13.0
GC: G1 Young Generation, G1 Old Generation
Memory: 4096M
Cores: 10
Metal Rendering is ON
Registry:
    analyze.exceptions.on.the.fly=true
    external.system.auto.import.disabled=true
    ide.text.editor.with.preview.show.floating.toolbar=false
    ide.instant.shutdown=false

Non-Bundled Plugins:
    wu.seal.tool.jsontokotlin (3.7.4)
    some.awesome (1.14)
    de.netnexus.camelcaseplugin (3.0.12)
    google-java-format (1.15.0.0)
    com.github.setial (4.0.2)
    com.github.copilot (1.1.34.2009)
    com.facebook.ktfmt_idea_plugin (1.1.0.41)
    VisualVMLauncher (1.21.211.6085.0)
    Gradle View (4.0.0)
    com.chuntung.plugin.gistsnippet (1.1.0)
    zielu.gittoolbox (212.9.9)
    com.developerphil.adbidea (1.6.8)
    com.thelumiereguy.compose_helper (2.0.3)

```

**Comment by pazlavi** [comment]
@cshmerling, 
I'm trying to retrieve the `AdId` and `AppSetId`, but I keep getting exceptions. 
```java
Caller is not authorized to call this API. Caller is not allowed.
```

![Screen Shot 2022-11-30 at 16 48 23](https://user-images.githubusercontent.com/72502546/204826985-cceda163-1083-42f1-aea4-c33c191a3403.png)


I'm using Canary 8 to run my app.

**Comment by cshmerling** [comment]
@pazlavi assuming you did everything as described in [the release notes about this](https://developer.android.com/design-for-safety/privacy-sandbox/release-notes#current), we did find out there were some additional permissions being checked to enable access to Privacy Sandbox APIs. We updated the [instructions](https://developer.android.com/design-for-safety/privacy-sandbox/setup-api-access#enable-access) specifically mentioning enabling the Sandbox via the UI:
```
adb shell am start -n com.google.android.adservices.api/com.android.adservices.ui.settings.AdServicesSettingsActivity
```

Can you confirm that you have done all that? If the issue still persists, please capture an `adb bugreport` and share it with us via our [IssueTracker](https://issuetracker.google.com/issues/new?component=1116743&template=1642602)

**Comment by pazlavi** [comment]
@cshmerling 
Unfortunately, it didn't help. I opened a bug report with `adb bugreport` output on Issue Tracker

**Comment by cshmerling** [comment]
Thanks for the detailed issue. I'm forwarding it to the right people to take a look.
The only other thing I might suggest is after enabling the Privacy Sandbox would be to reboot the device before trying to run the app as that has resolved some other issues, but you can also just wait for a response to the bug you filed as well

**Comment by pazlavi** [comment]
Thanks @cshmerling  for forwarding the issue. 
Rebooting the device didn't help here.
I'm curious why the issue is already closed as `won't fix` as it is clear that the `SecurityException` is thrown from your internal `AdServicesStatusUtils` class.

**Comment by cshmerling** [comment]
I think some wires got crossed by our support team- I've reopened the issue and we will look in to it

**Comment by cshmerling** [comment]
It looks like there were several issues- one permission issue and once that was resolved a new issue arose that the AdId returned was not useable. For the latter issue, we confirmed that these are known limitations. The getAdId and getAppSetId APIs do not work properly yet as we have not fully completed the required rollouts. Please stay tuned.

**Comment by cshmerling** [comment]
Given that this is a known limitation, I am going to close this issue. We will update our release notes when this functionality becomes available.

**Comment by yuvalcar** [comment]
Hey, 
Do you happen to have any news regarding this issue? 
Updating that it reproduces consistently for me when using the [sampleApp](https://github.com/android/privacy-sandbox-samples/tree/dev-preview-main/AttributionReporting/MeasurementSampleApp) locally. 

I'm using a pixel4 emulator with TiramisuPrivacySandbox SDK revision 9, running with arm image. 
When running the `registerSource` or `registerTrigger` methods, I'm getting the following: 
```
D/adservices: registerSource
E/adservices: Failed to get Ad ID: java.lang.SecurityException: Caller is not authorized to call this API. Caller is not allowed.
D/adservices: registerTrigger
E/adservices: Failed to get Ad ID: java.lang.SecurityException: Caller is not authorized to call this API. Caller is not allowed.
```

I'm using the following permissions in the app manifest: 
```
<uses-permission android:name="android.permission.INTERNET" />
  <uses-permission android:name="android.permission.ACCESS_ADSERVICES_ATTRIBUTION" />
  <uses-permission android:name="android.permission.ACCESS_ADSERVICES_AUDIENCES" />
  <uses-permission android:name="android.permission.ACCESS_ADSERVICES_TOPICS" />
  <uses-permission android:name="android.permission.ACCESS_ADSERVICES_AD_ID" />
  <uses-permission android:name="com.google.android.gms.permission.AD_ID" />
```

And I have the following ad_services_config.xml file configured: 
```
<ad-services-config>
    <ad-id allowAllToAccess="true"/>
    <attribution allowAllToAccess="true" />
    <includes-sdk-library allowAllToAccess="true" />
    <topics allowAllToAccess="true" />
    <custom-audiences allowAllToAccess="true" />
</ad-services-config>
```

and the following configs are set on the emulator device_config: 
```
adid_kill_switch=false
adservice_enable_status=true
adservice_enabled=true
adservice_system_service_enabled=true
disable_sdk_sandbox=false
global_kill_switch=false
ppapi_app_allow_list="*"
ppapi_app_signature_allow_list="*"
```

I even added our enrollment link to the `mdd_measurement_manifest_file_url config` config, although I assume it doesn't affect the local run. 

in a locally generated bugreport (which I can submit to the issue tracker if necessary) I can see the following log lines: 
```
./bugreport-sdk_gphone64_arm64-TRA7.230112.001.C4-2023-02-21-01-23-50.txt:69616:02-20 17:14:48.748 10162 12674 12674 D adservices: registerSource
./bugreport-sdk_gphone64_arm64-TRA7.230112.001.C4-2023-02-21-01-23-50.txt:69628:02-20 17:14:48.766 10150 12066 12738 E adservices: Failed to find resolveInfo for adServices service. Intent action: android.adservices.adid.AdIdProviderService
./bugreport-sdk_gphone64_arm64-TRA7.230112.001.C4-2023-02-21-01-23-50.txt:69629:02-20 17:14:48.766 10150 12066 12738 E adservices: Failed to find AdServices service
./bugreport-sdk_gphone64_arm64-TRA7.230112.001.C4-2023-02-21-01-23-50.txt:69630:02-20 17:14:48.768 10150 12066 12066 E adservices: Measurement API is disabled
./bugreport-sdk_gphone64_arm64-TRA7.230112.001.C4-2023-02-21-01-23-50.txt:69631:02-20 17:14:48.768 10150 12066 12066 E adservices: Measurement API is disabled
```

Will be happy to hear your thoughts on the matter @cshmerling :)

**Comment by erintwalsh** [comment]
Hi @yuvalcar, 

Apologies for the delay - are you still seeing this issue? 

If you are, 
`Caller is not allowed` generally means there is an issue with the allowlist. I would try deleting your current allowlist with `adb shell device_config delete adservices ppapi_app_allow_list` and entering your package name with `adb shell device_config put adservices ppapi_app_allow_list [package name]`. If you are using the Measurement sample app, you don't need to set the allowlist at all, as `com.example.measurement.sampleapp` is included by default. 

For `Measurement API is disabled`, try making sure the kill switches are off by running these commands: 
* `adb shell device_config put adservices adid_kill_switch false`
* `adb shell device_config put adservices measurement_kill_switch false`

Let us know if you're still seeing issues!

**Comment by notmariazoe** [comment]
Is this issue still outstanding? Please refer to the latest version of the Kotlin sample in the 'main' branch.

**Comment by notmariazoe** [comment]
Closing due to lack of activity.
